### PR TITLE
feat: map buses to balancing authorities

### PR DESCRIPTION
**Pull request by BainanXia**
[Pull Request doc](https://breakthrough-energy.github.io/docs/user/git_guide.html#d-pull-request)

### Purpose
This PR is created based on #180 for following reasons:
- The commit history of the original PR is hard to clean up. Discussed with @rouille , we decided to not bother and create a fresh PR.
- Address any remaining comments.

Implement a function to map buses to balancing authorities.

### What the code is doing
`map_buses_to_ba` is constructed by
- `map_buses_to_county` using census API to find the county that each query bus belongs to
- `BA_County_map.json` provides a list of counties that each BA belongs to, details of the file generation see `Dropbox (Gates Ventures)/BreakthroughEnergySciences/Explorations/BainanX/external connection requests/County to BA map`. There is a conversation regarding reformatting the file and the data source under the original [PR](https://github.com/Breakthrough-Energy/PreREISE/pull/180#discussion_r662575235)

### Testing
Tests for `map_buses_to_county` and `map_buses_to_county` are added. 

An integration test is conducted by inputing all buses in the base grid to the function, the result file is committed to this PR in the data folder as well for caching purpose, which takes a couple of hours to run. Note that there are 118 buses failed to be matched to a BA due to mismatch county names or outside the US territory, attaching the bus id here for reference:

```
[90,
 91,
 7991,
 7992,
 8706,
 8707,
 8708,
 18308,
 18309,
 18310,
 18311,
 18312,
 18349,
 18350,
 18351,
 18352,
 18353,
 18354,
 18355,
 18356,
 18357,
 18358,
 18359,
 18360,
 18388,
 18389,
 18390,
 19535,
 19536,
 19537,
 19538,
 25702,
 25703,
 40644,
 41013,
 41014,
 41015,
 41016,
 41017,
 70001,
 70003,
 70004,
 70005,
 70006,
 70007,
 70008,
 70009,
 70010,
 70011,
 70012,
 70013,
 70014,
 70015,
 70016,
 70017,
 70018,
 70019,
 70020,
 70021,
 70022,
 70023,
 70024,
 70025,
 70026,
 70027,
 70028,
 70029,
 70030,
 70031,
 70032,
 70033,
 70034,
 70035,
 70036,
 70037,
 70038,
 70039,
 70040,
 70041,
 70042,
 70043,
 70044,
 70045,
 70046,
 70047,
 2010201,
 2028491,
 2028492,
 2028705,
 2040000,
 2053120,
 2053174,
 2080030,
 2080031,
 2090000,
 2090001,
 2090002,
 2090003,
 2090004,
 2090005,
 2090006,
 2090007,
 2090008,
 2090009,
 2090010,
 2090011,
 2090012,
 2090013,
 2090014,
 2090015,
 2090016,
 2090017,
 2090018,
 2090019,
 2090020,
 2090021,
 2090022,
 2090023]
```

### Where to look
```
prereise/gather/demanddata/eia/map_ba.py
prereise/gather/demanddata/eia/tests/test_map_ba.py
prereise/gather/data/BA_County_map.json
```
Note that based on the observation, we have to loop through the bus entries via census API which gives the computational bottle neck of this function. It takes a couple of hours to get such mapping for 82k buses. I would suggest to cache the resultant mapping in the data folder for immediate usage.

### Time estimate
15 min, longer if one would like to track the source of the data.

### CODE REVIEW

**Comment by BainanXia** [comment]
Update: per @danielolsen 's feedback, it is worthwhile to visualize the buses that are not matched to any BA:
- Outside US territory
- County name mismatch from census API and BA_County_map.json

image

Note that
- The scatter plot is color coded by load zones
- The original Bokeh map locates in `Dropbox (Gates Ventures)/BreakthroughEnergySciences/Explorations/BainanX/external connection requests/updated all bus to BA map with nan`, with hover presenting county names and GIS of buses.

**Comment by danielolsen** [comment]
Lots of new buses we added for offshore wind!

**Comment by danielolsen** [comment]
For the counties that are not present in **BA_County_map.json** and result in NaN values for that reason, can we augment this data file to fix the situation? Looking at your previous county-to-BA map, the three counties in Virginia should be PJM, and the county in Georgia should be SOCO.

Then all remaining coordinates should be ones that are outside the borders of the U.S., and we can use geographical or topological proximity to automatically address these. Right?

**Comment by BainanXia** [comment]
> For the counties that are not present in **BA_County_map.json** and result in NaN values for that reason, can we augment this data file to fix the situation? Looking at your previous county-to-BA map, the three counties in Virginia should be PJM, and the county in Georgia should be SOCO.
> 
> Then all remaining coordinates should be ones that are outside the borders of the U.S., and we can use geographical or topological proximity to automatically address these. Right?

Yes, I also updated my comment above. Then the question is should this process be in the code base, if so, should it be part of this PR.

**Comment by danielolsen** [comment]
I think it is reasonable to add the nearest-neighbor algorithm within the codebase and to do so within this PR. We already import PowerSimData, so we have access to `powersimdata.utility.distance.haversine`, and I think we can use pandas `map`/`apply` to do this fairly simply.

**Comment by rouille** [comment]
You can use the angular_distance function in the distance module to calculate the nearest neighbor.

**Comment by BainanXia** [comment]
@danielolsen @rouille A summary of the latest update:
- For county name mismatch between Census API and `BA_County_map.json`, I put the manual fix in the function and update the `BA_County_map.json` (some automations added in the process of generating this file), together with these to fixes, we eliminate all buses within the US territory.
- For buses outside the US territory, I implement the nearest neighbor logic as discussed above taking advantages of `find_nearest_neighbor` function in the `distance.py` module.
- Now the `map_buses_to_ba` function returns `bus_ba_map` and `bus_no_county_match`, instead of `bus_ba_map` and `bus_no_ba_match`.
- The cache file is updated to the latest, i.e. all buses are assigned to BA in the base grid, note that for the buses outside the US territory, we only assign BAs to them, the County field remains to be None for reference purpose.

File: prereise/gather/demanddata/eia/map_ba.py [hunk 1]
```diff
@@ -95,3 +102,56 @@ def map_buses_to_county(bus_county_map):
...
+
+    bus_ba_map, bus_no_county_match = map_buses_to_county(bus_df)
+    # hard coded fix for county name mismatch between the returns from census API and
+    # BA_County_map.json
+    bus_ba_map.County.replace(
+        {
+            "Danville City__VA": "Danville__VA",
+            "Harrisonburg City__VA": "Harrisonburg__VA",
+            "Franklin City__VA": "Franklin__VA",
+        },
```
Comment (danielolsen) [inline prereise/gather/demanddata/eia/map_ba.py:127]: Do you think it is worth putting this in a **const.py** file?

Comment (BainanXia) [reply prereise/gather/demanddata/eia/map_ba.py:127]: Given it doesn't have a concrete meaning outside this function, I would suggest to keep it here.

**Review by danielolsen — APPROVED** [review]

**Pull request merged**